spark3/src/main/java/org/apache/iceberg/spark/source/SparkMergeScan.java

```diff
@@ -0,0 +1,178 @@
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.spark.Spark3Util;
+import org.apache.iceberg.util.PropertyUtil;
+import org.apache.iceberg.util.TableScanUtil;
+import org.apache.spark.broadcast.Broadcast;
+import org.apache.spark.sql.connector.read.Statistics;
+import org.apache.spark.sql.connector.read.SupportsReportStatistics;
+import org.apache.spark.sql.util.CaseInsensitiveStringMap;
+
+import static org.apache.iceberg.TableProperties.SPLIT_LOOKBACK;
+import static org.apache.iceberg.TableProperties.SPLIT_LOOKBACK_DEFAULT;
+import static org.apache.iceberg.TableProperties.SPLIT_OPEN_FILE_COST;
+import static org.apache.iceberg.TableProperties.SPLIT_OPEN_FILE_COST_DEFAULT;
+import static org.apache.iceberg.TableProperties.SPLIT_SIZE;
+import static org.apache.iceberg.TableProperties.SPLIT_SIZE_DEFAULT;
+
+class SparkMergeScan extends SparkBatchScan implements SupportsReportStatistics {
+
+  private final Table table;
+  private final boolean ignoreResiduals;
+  private final Schema expectedSchema;
+  private final Long snapshotId;
+  private final Long splitSize;
+  private final Integer splitLookback;
+  private final Long splitOpenFileCost;
+
+  // lazy variables
+  private List<FileScanTask> files = null; // lazy cache of files
+  private List<CombinedScanTask> tasks = null; // lazy cache of tasks
+
+  SparkMergeScan(Table table, Broadcast<FileIO> io, Broadcast<EncryptionManager> encryption,
+                 boolean caseSensitive, boolean ignoreResiduals, Schema expectedSchema,
+                 List<Expression> filters, CaseInsensitiveStringMap options) {
+
+    super(table, io, encryption, caseSensitive, expectedSchema, filters, options);
+
+    this.table = table;
+    this.ignoreResiduals = ignoreResiduals;
+    this.expectedSchema = expectedSchema;
+    this.snapshotId = Spark3Util.propertyAsLong(options, "snapshot-id", null);
+
+    Map<String, String> props = table.properties();
+
+    long tableSplitSize = PropertyUtil.propertyAsLong(props, SPLIT_SIZE, SPLIT_SIZE_DEFAULT);
+    this.splitSize = Spark3Util.propertyAsLong(options, "split-size", tableSplitSize);
+
+    int tableSplitLookback = PropertyUtil.propertyAsInt(props, SPLIT_LOOKBACK, SPLIT_LOOKBACK_DEFAULT);
+    this.splitLookback = Spark3Util.propertyAsInt(options, "lookback", tableSplitLookback);
+
+    long tableOpenFileCost = PropertyUtil.propertyAsLong(props, SPLIT_OPEN_FILE_COST, SPLIT_OPEN_FILE_COST_DEFAULT);
+    this.splitOpenFileCost = Spark3Util.propertyAsLong(options, "file-open-cost", tableOpenFileCost);
+
+    if (snapshotId == null) {
+      // init files with an empty list if the snapshot id is not set to avoid picking any concurrent changes
+      files = Collections.emptyList();
+    }
+  }
+
+  Long snapshotId() {
+    return snapshotId;
+  }
+
+  @Override
+  public Statistics estimateStatistics() {
+    if (snapshotId == null) {
+      return new Stats(0L, 0L);
```

Comment:
Why would `snapshotId` be null? Don't we need to fix the snapshot ID so that we can use it as the base for validations in the replace commit?

Reply:
We do fix the snapshot id and make sure we use it consistently. It can be null if we started with an empty table.

Reply:
Makes sense. I always forget about empty tables.